ctx.go

```diff
@@ -1054,6 +1055,31 @@ func (c *Ctx) Query(key string, defaultValue ...string) string {
 	return defaultString(c.app.getString(c.fasthttp.QueryArgs().Peek(key)), defaultValue)
 }
 
+// Queries returns a map of query parameters and their values.
+// It returns an error if the query string is not valid.
+func (c *Ctx) Queries() (map[string]string, error) {
+	queryRaw := c.fasthttp.QueryArgs().String()
+	queryRaw, err := url.QueryUnescape(queryRaw)
+	if err != nil {
```

Comment:
If the string is already escaped this will fail. It can be tested by adding a test that sets URLEscaped values

Reply:
> If the string is already escaped this will fail. It can be tested by adding a test that sets URLEscaped values

Do you mean I do this?
```go
queryRaw := c.fasthttp.QueryArgs().String()

queryRaw, err := url.QueryUnescape(queryRaw)
if err != nil {
    queryRaw = url.QueryEscape(queryRaw)
}
```

Reply:
I think you mean to add tests for failing the `Queries` function.
I added it too.

Reply:
why so complicated ?

```go
m := make(map[string]string, c.Context().QueryArgs().Len())
c.Context().QueryArgs().VisitAll(func(key, value []byte) {
	m[string(key)] = string(value)
})
```
don't really know if we need this functionality for the 1 line either
otherwise we would have to provide a function for each of them
-> form values, query values, route params

Reply:
> why so complicated ?
> 
> ```go
> m := make(map[string]string, c.Context().QueryArgs().Len())
> c.Context().QueryArgs().VisitAll(func(key, value []byte) {
> 	m[string(key)] = string(value)
> })
> ```
> 
> don't really know if we need this functionality for the 1 line either otherwise we would have to provide a function for each of them -> form values, query values, route params

Your right, But this code can't parse this query encoded URL: `id%3D1%26name%3Dtom%26hobby%3Dbasketball%2Cfootball%26favouriteDrinks%3Dmilo%2Ccoke%2Cpepsi%26alloc%3D%26no%3D1`

In this problem, we should unescape these encoded queries. Would you happen to have any suggestions?

Reply:
why should we parse the encodede url? the reason for encoding is that these control characters are no longer interpreted as such

Reply:
correct because it is invalid to process encoded query strings functionally
when transferring via http you have to stick to the standard, if you remove the control characters by encoding, then the not possible analysis of the string is exactly what the person wanted to achieve

Reply:
> correct because it is invalid to process encoded query strings functionally when transferring via http you have to stick to the standard, if you remove the control characters by encoding, then the not possible analysis of the string is exactly what the person wanted to achieve

Your right.